HIVE-26986: Fix OperatorGraph when a query plan contains UnionOperator

ngsg opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
I rewrite OperatorGraph to make sure that OperatorGraph represents actual Tez DAG.
In new OperatorGraph, each cluster has a unique root operator and one Operator can belong to more than one cluster.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Tez DAG and current OperatorGraph group operators differently when a query plan contains union operators. More precisely, OperatorGraph groups all root operators of a union operator into the same cluster, but Tez DAG creates a BaseWork for each of the root operators and clones descendants of the union operator if needed. This difference makes false-positive errors when detecting parallel edges, which leads to insertion of unnecessary concentrator RS by ParallelEdgeFixer.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

I run the below maven tests and add a unit test for OperatorGraph.
```
mvn test -Dtest=TestSharedWorkOptimizer
cd itests; mvn test -Dtest=TestCliDriver
```

deniskuzZ commented:
@ngsg, could you please reopen this PR, can't do it from UI

ngsg commented:
@deniskuzZ , it seems that I don't have a permission to reopen this PR. I rebased and force-pushed my branch recently so there was a difference between the PR's head and that of my branch. I restore my branch and now they are pointing the same commit. Could you please check whether you can reopen this PR? If you can't, I'll create a new PR based on the latest commit on master branch.

ngsg commented:
@deniskuzZ , I rebased the branch.

aturoczy commented:
Can somebody retrigger pls? The test failed due to a testCLIDriver issue. I do not think it is related

okumin commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/graph/OperatorGraph.java:
```diff
@@ -81,183 +76,202 @@ public EdgeType getEdgeType() {
+      clusters.add(createCluster(rootOperator));
+    }
+
+    for (Operator<?> operator: mergeJoinOperators) {
```
Why do we need to merge clusters with CommonMergeJoinOperators?

ngsg replied:
@okumin, sorry for the late response.
An unmerged cluster is a group of operators that are descendents of the same operator. CommonMergeJoinOperator can have multiple ReduceSink ancestors, so one CommonMergeJoinOperator belongs to multiple unmerged clusters at this step. To prevent this issue, we have to merge these clusters and make a unique cluster for a CommonMergeJoinOperator.

ngsg replied:
We need this for-loop because not all MergeJoinOperators are root operators of vertices. For example, you can see that the MergeJoin operator is a descendant of another operator in the following qfile output:
https://github.com/apache/hive/blob/e8fc30c95a817b0c3ef0cc2c4c63a2fd63b2fcee/ql/src/test/results/clientpositive/llap/auto_sortmerge_join_12.q.out#L318-L332

okumin replied:
The example explains the answer to my question well. Thanks.

okumin commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java:
```diff
@@ -1863,12 +1863,19 @@ private static boolean validPreConditions(ParseContext pctx, SharedWorkOptimizer
+        Set<Cluster> cc2 = cluster2.childClusters(edgePredicate);
+
+        if (!Collections.disjoint(cc1, cc2)) {
+          LOG.debug("merge would create an unsupported parallel edge(CHILDS)", op1, op2);
```
nit: We may want to debug op1 and op2

ngsg replied:
Thanks for pointing that out; fixed debug messages

okumin reviewed (commented):
This seems to be working for some test cases where I took a look. I grabbed the overview but I'm not yet confident about this change.

deniskuzZ commented:
hi @okumin, do you recall if there was a decision on that PR?

zabetak commented:
@deniskuzZ From the JIRA discussion it seems like a PR that we would like to merge. Unfortunately, it was stuck and got stale due to lacking reviews.

deniskuzZ commented:
let me reopen it :) yes, we are very short on reviewers

okumin commented:
I can take a look. It might take a couple of days

okumin commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java:
```diff
@@ -185,23 +188,36 @@ private String sig(Pair<Operator<?>, Operator<?>> o1) {
   }
 
+  private static class ActualEdgePredicate implements OperatorGraph.OperatorEdgePredicate {
+    EnumSet<EdgeType> acceptableEdgeTypes = EnumSet.of(EdgeType.FLOW, EdgeType.SEMIJOIN, EdgeType.BROADCAST);
```
Don't we need DPP?

ngsg replied:
We don't need DPP here because DPP does not make a Tez edge; the dynamic partitioning information is sent directly from `AppMasterEventOperator` to target `TableScanOperator` via Tez event API.

okumin replied:
That's true. Please make make it private static constant variable.

ngsg commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/graph/OperatorGraph.java:
```diff
@@ -81,183 +76,202 @@ public EdgeType getEdgeType() {
+    return clusters;
+  }
+
+  private DagGraph<Operator<?>, OperatorGraph.OpEdge> createDagGraph(ParseContext pctx) {
```
Apologies if I misunderstood your review. I think the current `DagGraph` is identical to the previous `DagGraph`. The only difference is that I removed the `index` field in the `OpEdge` class since it was not being used.

ngsg commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/graph/OperatorGraph.java:
```diff
@@ -81,183 +76,202 @@ public EdgeType getEdgeType() {
+            MapJoinOperator childMJOperator = (MapJoinOperator) childOperator;
+            int parentTag = childMJOperator.getParentOperators().indexOf(operator);
+            int bigTablePos = childMJOperator.getConf().getPosBigTable();
+            if (parentTag == bigTablePos) {
```
### If it is a pure Map Join
Current implementation does not create duplicated cluster: if the `childMJOperator` is a pure MapJoin, then the if-condition `parentTag == bigTablePos` is always false because there is no RS between the `childMJOperator` and big-table-side TableScanOp. In fact, the if-condition is used to prevent a pure MapJoin from becoming a root operator, but I think the current implementation is non-trivial and hard to understand.
I'll update the if-clause to use `isDynamicPartitionHashJoin()` instead of `parentTag == bigTablePos`.

### If it is a DPHJ
You're right—DPHJ and MergeJoin behave similarly. I'll change `mergeJoinOperators` to `joinOperators`, which will  contain both MergeJoin and DPHJ.

ngsg commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/graph/OperatorGraph.java:
```diff
@@ -81,183 +76,202 @@ public EdgeType getEdgeType() {
-      if (currentCluster == null) {
-        currentCluster=new Cluster();
-        currentCluster.add(curr);
+  private Set<Cluster> createClusterSet(ParseContext pctx) {
```
You're right.
The original implementation assumes that every operator belongs to a single-merged cluster, so it uses `nodeCluster: Map[Operator, Cluster]`. However, the assumption is incorrect.

okumin reviewed (commented):
@ngsg Thanks for simplifying some logic. I'm more confident now. I'm about to merge this PR.

I have two requests.
1. I left one minor cosmetic change request. Please take a look
2. Some qtests failed probably because they were added recently. Please regenerate q.out files

ngsg commented:
@okumin, thank you for your review. I have checked and addressed your requests. Could you kindly review the changes once more? Thank you.

okumin commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java:
```diff
@@ -185,23 +187,37 @@ private String sig(Pair<Operator<?>, Operator<?>> o1) {
   }
 
+  private static class ActualEdgePredicate implements OperatorGraph.OperatorEdgePredicate {
+    private static final EnumSet<EdgeType> acceptableEdgeTypes =
```
```suggestion
    private static final EnumSet<EdgeType> ACCEPTABLE_EDGE_TYPES =
```

ngsg replied:
Modified the code as suggested.

okumin commented:
I am waiting for CI, adding one comment

okumin commented:
I still see at least one related error.
https://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3998/4/tests

When you update it, it would be appreciated if you could follow some advice from SonarCloud. "Cognitive Complexity" and "Braint Method" are almost impossible to fix, but some others might be possible. Only easy ones related to this PR are ok.
https://sonarcloud.io/project/issues?id=apache_hive&pullRequest=3998&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

ngsg commented:
I have updated the missing qtest output and addressed some issues reported by SonarCloud.
However, I haven't implemented "usage of generic wildcard type" in OperatorGraph.java as suggested because resolving this issue would likely require either extensive refactoring or double unchecked type casting, which is another type of code smell.
(e.g. `return (DagGraph<Operator<OperatorDesc>, OpEdge>) (DagGraph<?, OpEdge>) dagGraph;`)
If you have any alternative suggestions, I would appreciate your input.

okumin reviewed (approved):
+1. I retriggered CI as it failed due to probably an unrelated error.
